Project 4/README.md

```diff
@@ -45,3 +45,212 @@ You're probably thinking, "why would I submit a level 8 kata if they're not wort
+"prices" of mile-tickets are much lower than the typical prices, given 100 member mileages is worth 1 dollar. Mike also learns that some credit card 
+companies offer initial bonus, which typically are credit card points. Beside, some credit cards have such a benefit that the credit card points can be
+transferred to mileages of some air companies, and the ratio is 1:1, which means one point can be transferred as one mileage. So Mike thinks
+that applying for a specific credit card is a way to save money to buy flight tickets. There is already an ever-updating dataset of credit cards offers 
+(http://creditcardoffers.org/), so he wants to use the dataset and makes a SPARQL query, called "credit card offer" to look for relevant credit card 
+offers. 
+
+Description of the offers (no more than 5 offers):
+1) The credit card points can be transfer to one of UA, Delta, AA.
+2) The initial bonus is more than 60,000 points.
+3) The annual fee is less than 250 dollars.  
+4) The spending requirement for the initial bonus is no more than 4000 dollars in the first three month of the account openning. 
+
+``` 
+PREFIX cdt: < http://creditcardoffers.org/type/ >
+PREFIX cdo: < http://creditcardoffers.org/object/ >
+
+SELECT ?cardName ?airCom ?iniPoints ?annFee ?spending 
+
+WHERE{ 
+   ?item cdt:label ?cardName;
+         cdt:transfer_mile ?airCom.
+	FILTER (str(?airCom)= "UA","AA","Delta")
+	 cdt:initial_bonus ?iniPoints
+	FILTER (?iniPoints > 60000)
+	 cdt: annual_fee ?annFee
+	FILTER (?annFee <= 200)
+	 cdt:spending_req ?spending
+	FILTER (?spending <= 3000)
+}
+LIMIT 5
+
+
+```
+Query 3. (Kata 4 -10 points)
+
+Title: the politicians from politician families in China.
+
+Description: 
+
+A research team, studying Chinese electoral system, wants to know how many Chinese politician are from politician families. Based on their initial 
+investigation, Chinese electoral system is not as stated in the Chinese constitution, where the government leaders at all levels are elected by the
+people. 
+
+A politician P from politician family is defined as below:
+1) One of P's biological parent is or was a politician; or
+2) One of P's legal parent is or was a politician; or
+3) One of P's spouse or ex-spouse is a politician; or
+4) One of P's spouse's biological parent in law is or was a politician; or
+5) One of P's spouse's legal parent is or was a politican; or
+6) P is a Chinese citizen; and
+7) P was born after 1949.
+
+```
+
+PREFIX wd: <http://www.wikidata.org/entity/>
+PREFIX wdt: <http://www.wikidata.org/prop/direct/>
+PREFIX wikibase: <http://wikiba.se/ontology#>
+
+# politician from family
+
+SELECT (COUNT (?item) AS ?numPoliticianFromFamily) (COUNT (?politician) AS ?totalPolitician)
+
+WHERE {
+  
+  ?item wdt:P31 wd:Q5.          #instance of human.
+  ?item wdt:P27 wd:Q148.        # a citizen of PR of China.  
+  ?item wdt:P106 wd:Q82955.      # politician.
+  {?item wdt:P22 ?father.       
```

Comment:
Use the VALUES keyword instead of UNION here. UNION increases the computational complexity of the query significantly, VALUES doesn't. 
Otherwise, I'm liking this kata :)

Reply:
Hi John,
Can VALUES keywords work for predicate? I plan to make the object as constant, such as wd:Q82955, and the predicate a variable. For example, ?item ?relation wd:Q82955 #as politician, and then use VALUES keywords for ?relation, qualifying the relations by father, mother, spouse, etc. But I tried, and it does not work.
Bests,
Jieming

Reply:
@Jiemingy Yep! For example, instead of: 
```
  UNION {?item wdt:P25 ?mother.
               ?mother wdt:P106 wd:Q82955.}   
  UNION {?item wdt:P8810 ?parent.
               ?parent wdt:P106 wd:Q82955.} 
```
You can write: 
```
VALUES ?predicate (wdt:P25 wdt:P8810)
SELECT...WHERE...
    ?item ?predicate ?momordad .
    ?momordad wdt:P106 wd:Q82955.
```

Reply:
Great! Thank you!
_Jieming